GSoC 2021 Project: Add regexp-parser()

LittleFish33 opened the pull request:
[GSoC2021-Idea-&-Project-list#add-regexp-parser](https://github.com/syslog-ng/syslog-ng/wiki/GSoC2021-Idea-&-Project-list#add-regexp-parser)

This PR introduces a parser that can parse messages with regular expressions.

## Options of regexp parsers:
### Declaration:
```
parser parser_name {
  regexp-parser(
    patterns()
    prefix()
    flags()
  );
};
```
### patterns()
regular expression patterns that you want to find a match, regexp-parser supports multiple patterns
### prefix()
Insert a prefix before the name part of the parsed name-value pairs to help further processing
### flags()
Synopsis: disable-jit, dupnames, global, ignore-case, newline, store-matches, unicode, utf8

### Example
```
parser {
  regexp-parser(
    prefix(".regexp.") 
    patterns("(?<DN>foo)", "(?<DN>ball)")
    flags(ignore-case)
    template("${MESSAGE}")
  );
};
```

kira-syslogng commented:
This user does not have permission to start the build. Can one of the admins verify this patch and start the build?
(admin: you have the next options (make sure you checked the code):
"ok to test" to accept this pull request (and further changes) for testing
"test this please" for a one time test run
"add to whitelist" add author of a Pull Request to whitelist (globally, be careful, it means this user can trigger kira for any PR)
do nothing -> CI won't start)

kira-syslogng commented:
This user does not have permission to start the build. Can one of the admins verify this patch and start the build?
(admin: you have the next options (make sure you checked the code):
"ok to test" to accept this pull request (and further changes) for testing
"test this please" for a one time test run
"add to whitelist" add author of a Pull Request to whitelist (globally, be careful, it means this user can trigger kira for any PR)
do nothing -> CI won't start)

bazsi commented:
@kira-syslogng  ok to test

kira-syslogng commented:
Build SUCCESS

bazsi commented on lib/logmatcher.h:
```diff
@@ -66,6 +66,7 @@ struct _LogMatcher
   gint ref_cnt;
   gint flags;
   gchar *pattern;
+  gchar *prefix;
```
This should go in the PCRE specific matcher class, that is defined in logmatcher.c and is called LogMatcherPcreRe

LittleFish33 replied:
Thanks! But I have a question. If I define the `prefix` in the `LogMatcherPcreRe` class, I can't set the prefix in the `regexp-parser.c` because the definition of `LogMatcherPcreRe` is in the `logmatcher.c`. I can think of two ways to solve this problem:

1. move the definition of `LogMatcherPcreRe` from `logmatcher.c` to `logmatcher.h` (but may be I should also move the definitions of `LogMatcherGlob` and `LogMatcherString` to  `logmatcher.h` for consistency)
2. declare a new function called `log_matcher_pcre_set_prefix` in `logmatcher.h`, which can set the prefix of a LogMatcherPcreRe object

bazsi commented on lib/logmatcher.c:
```diff
@@ -209,6 +209,7 @@ log_matcher_string_new(const LogMatcherOptions *options)
   self->super.compile = log_matcher_string_compile;
   self->super.match = log_matcher_string_match;
   self->super.replace = log_matcher_string_replace;
+  self->super.prefix = NULL;
```
this is not needed if https://github.com/syslog-ng/syslog-ng/pull/3702/files#r650710436 is fixed.

bazsi commented on lib/logmatcher.c:
```diff
@@ -279,6 +280,7 @@ log_matcher_glob_new(const LogMatcherOptions *options)
   self->super.match = log_matcher_glob_match;
   self->super.replace = NULL;
   self->super.free_fn = log_matcher_glob_free;
+  self->super.prefix = NULL;
```
likewise.

bazsi commented on modules/Makefile.am:
```diff
@@ -47,6 +47,7 @@ include modules/systemd-journal/Makefile.am
 include modules/system-source/Makefile.am
 include modules/tagsparser/Makefile.am
 include modules/xml/Makefile.am
+include modules/regexp/Makefile.am
```
maybe we should rename the directory to regexp-parser, regexps are used in a lot of ways within syslog-ng and not all of them would concentrate in the "modules/regexp" directory (neither now or long term), so let's try to use a more specific directory.

bazsi commented on lib/logmatcher.c:
```diff
@@ -445,7 +447,17 @@ log_matcher_pcre_re_feed_named_substrings(LogMatcher *s, LogMessage *msg, int *m
             continue;
 
-          log_msg_set_value_by_name(msg, tabptr + 2, value + begin_index, end_index - begin_index);
+          if(self->super.prefix != NULL)
+            {
+              GString *formatted_key = g_string_new(self->super.prefix);
+              g_string_append(formatted_key, tabptr + 2);
+              log_msg_set_value_by_name(msg, formatted_key->str, value + begin_index, end_index - begin_index);
+              g_string_free(formatted_key, TRUE);
+            }
+          else
+            {
+              log_msg_set_value_by_name(msg, tabptr + 2, value + begin_index, end_index - begin_index);
+            }
```
This can be a pretty hot path so we need to be aware of (a) performance implications. Also, this code is becoming a bit too complex, so (b) extracting the functionality into a separate function seems to be a good move.

Hints:
(a) g_string_new() is pretty slow, it allocates both a GString header and then the payload, which is potentially reallocated right at the next line. So we have 2-3 memory allocations plus a free operation in a path that is executed multiple times for each message. Message rates can be pretty high with syslog-ng, reaching 100s of thousands per second, although regexps would slow those down to a 10s of thousands. In any case, this is potentially executing 100k per sec. There's an API on top of GStrings that will give you cheap/free allocation of GString instances, look at ```scratch-buffers.h``` and try to find a user of that code for examples.

(b) you could move this part in its entirety into a local/static function, which would make this function a lot easier to read.

bazsi commented on modules/regexp/regexp-parser-grammar.ym:
```diff
@@ -0,0 +1,87 @@
+parser_regexp_opt
+	: { last_matcher_options = regexp_parser_get_matcher_options(last_parser); } matcher_option
+	| KW_PREFIX '(' string ')'		{ regexp_parser_set_prefix(last_parser, $3); free($3); }
+    | KW_PATTERN '(' string ')'		{ regexp_parser_set_pattern(last_parser, $3); free($3); }
```
we should support multiple patterns, just like date-parser() does that with multiple potential formats. If any of the pattern matches we should accept (match) the message.

If none of them matches we should just return "unmatched", just as date-parser() does this when none of the formats match.

bazsi commented on modules/regexp/regexp-parser.c:
```diff
@@ -0,0 +1,175 @@
+  return log_matcher_compile(self->matcher, self->pattern, NULL);
+}
+
+static const gchar *
+_get_formatted_key_with_prefix(RegexpParser *self, const gchar *key, GString *formatted_key)
+{
+  if (formatted_key->len > 0)
+    g_string_truncate(formatted_key, self->prefix_len);
+  else
+    g_string_assign(formatted_key, self->prefix);
+  g_string_append(formatted_key, key);
+  return formatted_key->str;
+}
+
+static const gchar *
+_get_formatted_key(RegexpParser *self, const gchar *key, GString *formatted_key)
+{
+  if (!self->prefix)
+    return key;
+  return _get_formatted_key_with_prefix(self, key, formatted_key);
+}
```
these functions do not seem to be used.

bazsi commented on lib/logmatcher.h:
```diff
@@ -66,6 +66,7 @@ struct _LogMatcher
   gint ref_cnt;
   gint flags;
   gchar *pattern;
+  gchar *prefix;
```
I can't see how this prefix gets initialized.

bazsi commented:
On Mon, Jun 14, 2021 at 5:13 PM LittleFish ***@***.***> wrote:

> *@LittleFish33* commented on this pull request.
> ------------------------------
>
> In lib/logmatcher.h
> <https://github.com/syslog-ng/syslog-ng/pull/3702#discussion_r651039461>:
>
> > @@ -66,6 +66,7 @@ struct _LogMatcher
>    gint ref_cnt;
>    gint flags;
>    gchar *pattern;
> +  gchar *prefix;
>
> Thanks! But I have a question. If I define the prefix in the
> LogMatcherPcreRe class, I can't set the prefix in the regexp-parser.c
> because the definition of LogMatcherPcreRe is in the logmatcher.c. I can
> think of two ways to solve this problem:
>
>    1. move the definition of LogMatcherPcreRe from logmatcher.c to
>    logmatcher.h (but may be I should also move the definitions of
>    LogMatcherGlob and LogMatcherString to logmatcher.h for consistency)
>    2. declare a new function called log_matcher_pcre_set_prefix in
>    logmatcher.h, which can set the prefix of a LogMatcherPcreRe object
>
>
>
the 2nd looks good. we usually do a cast in the impementation function in
these cases and take a pointer to the parent class, as it is simpler in the
grammar, which uses a single variable to represent any LogMatcher instance.

-- 
Bazsi

kira-syslogng commented:
Build SUCCESS

bazsi commented on lib/logmatcher.c:
```diff
@@ -435,6 +458,7 @@ log_matcher_pcre_re_feed_named_substrings(LogMatcher *s, LogMessage *msg, int *m
       /* Now we can scan the table and, for each entry, print the number, the name,
          and the substring itself.
        */
+      GString *formatted_name = scratch_buffers_alloc();
```
this allocation can move inside the log_matcher_pcre_re_feed_value_by_name() function, so you don't have to pass it as an argument.

bazsi commented on lib/logmatcher.c:
```diff
@@ -414,6 +417,26 @@ log_matcher_pcre_re_feed_backrefs(LogMatcher *s, LogMessage *msg, gint value_han
+  if(s->prefix != NULL)
+    {
+      if (formatted_name->len > 0)
+        g_string_truncate(formatted_name, s->prefix_len);
```
formatted_name can be a newly allocated GString instance, so truncating may not produce the required prefix. you will need to assign a new value every time.

bazsi replied:
I just recognized that there's a similar pattern in kv-parser(), which seems to be buggy as well. The scratch-buffers allocator does not guarantee that we get the same GString instance every time (though it may happen). Can you fix that part too?

LittleFish33 replied:
> formatted_name can be a newly allocated GString instance, so truncating may not produce the required prefix. you will need to assign a new value every time.

Hi, bazsi, I have a question. In the `scratch_buffers_alloc` function, the allocated buffer will be truncated to 0.
```c++
// lib/scratch-buffers.c, lines 127-128
GString *buffer = g_ptr_array_index(scratch_buffers, scratch_buffers_used);
g_string_truncate(buffer, 0);
```
In this case, if formatted_name is a newly allocated GString instance, it will go to the else case and call `g_string_assign(formatted_name, s->prefix);`.

bazsi replied:
you are right. I missed the loop on the regexp values, with scratch_buffers_alloc() before the loop it all makes sense now.

sorry.

bazsi commented on modules/regexp-parser/regexp-parser.c:
```diff
@@ -0,0 +1,170 @@
+      self->matchers = g_list_prepend(self->matchers, log_matcher_new(&self->matcher_options));
+      log_matcher_pcre_set_prefix((LogMatcher *)self->matchers->data, self->prefix);
+
+      if(!log_matcher_compile((LogMatcher *)self->matchers->data, item->data, NULL))
```
do we handle correctly the case where the 3rd pattern fails to compile? any memory leaks?

LittleFish33 replied:
@bazsi I'm a little confused about this problem. Does the memory leak here mean that the space allocated by `log_matcher_new` has not been released? But if any pattern fails to compile, the program will terminal directly and print the error on the console.

bazsi replied:
I just wanted to confirm that if the loop exits early (because of log_matcher_compile() returning failure), no memory leaks remain.

It is true that the exiting process would free all allocations, but we should ensure that all internal objects do that on their own without relying on the process to exit.

In any case, just make sure that log_matcher_free() is properly called for the partially populated list too.

bazsi commented on modules/regexp-parser/regexp-parser.c:
```diff
@@ -0,0 +1,170 @@
+                evt_tag_str ("input", input),
+                evt_tag_str ("pattern", ((LogMatcher *)item->data)->pattern));
+
+      if (log_matcher_match((LogMatcher *)item->data, *pmsg, LM_V_NONE, input, input_len))
```
although it's a bit of a layering violation, we may actually know that we are matching against LM_V_MESSAGE (in case the template() option is unset). And with that knowledge the LogMatcher instance can store indirect matches instead of full copies.

So instead of LM_V_NONE, you could use a conditional to see if ```self->super.template``` is set, and use LM_V_NONE if it is, and LM_V_MESSAGE if it's not.

This would result in the use of "indirect references" within the LogMessage, meaning that slices of the MESSAGE payload would be used as values. This might compress the size of the serialized log message and also reduce CPU usage as the new name-value pair is stored.

bazsi commented on modules/regexp-parser/regexp-parser.c:
```diff
@@ -0,0 +1,170 @@
+  if (self->matchers)
+    {
+      g_list_free(self->matchers);
+    }
```
g_list_free_full() allows you to pass a GDestroyNotify callback.

bazsi reviewed (commented):
This is good progress. A few more nitpicks added as comments. I was especially happy about the unit test.

Can you perhaps add a Python based functional test too? (tests/python_functional) it is a pytest based test suite. It's not a complete test suite yet, but it is the preferred place for functional tests.

You should find examples in that directory and @mitzkia can also help.

bazsi commented:
@LittleFish33 any plans on when you start working on my comments? thanks.

LittleFish33 commented:
@bazsi Sorry for the long delay, I will continue to work on this tomorrow

kira-syslogng commented:
Build SUCCESS

LittleFish33 commented:
@bazsi I have updated the commit and added a Python based functional test. Thanks for your comments.

bazsi commented:
I'll check it out in the coming days.


On Sat, Jul 3, 2021, 09:30 LittleFish ***@***.***> wrote:

> @bazsi <https://github.com/bazsi> I have updated the commit and added a
> Python based functional test. Thanks for your comments.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/syslog-ng/syslog-ng/pull/3702#issuecomment-873362478>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAFOK5TGF3OAQNWJBEHCS3TTV24BHANCNFSM46S7ZNWQ>
> .
>

bazsi commented:
Can you pls rebase to resolve the conflict? Thanks

kira-syslogng commented:
Build FAILURE

kira-syslogng commented:
Build FAILURE

kira-syslogng commented:
Build FAILURE

kira-syslogng commented:
Build FAILURE

kira-syslogng commented:
Build FAILURE

LittleFish33 commented:
I have resolved the conflict, but some checks were not successful. It shows that `Kira-starter — Build FAILURE`. But I just resolved the conflict by replacing
```
modules/(afamqp|affile|afmongodb|afprog|afsmtp|afsocket|afsql|afstomp|afstreams|afuser|azure-auth-header|basicfuncs|cef|confgen|cryptofuncs|csvparser|timestamp|diskq|dbparser|geoip2|graphite|json|kvformat|linux-kmsg-format|pacctformat|pseudofile|python|redis|riemann|syslogformat|systemd-journal|getent|system-source|stardate|snmptrapd-parser|xml|openbsd|examples|kafka|afsnmp|mqtt|[^/]*$)
```
with
```
modules/(afamqp|affile|afmongodb|afprog|afsmtp|afsocket|afsql|afstomp|afstreams|afuser|azure-auth-header|basicfuncs|cef|confgen|cryptofuncs|csvparser|timestamp|diskq|dbparser|geoip2|graphite|json|kvformat|linux-kmsg-format|pacctformat|pseudofile|python|redis|riemann|syslogformat|systemd-journal|getent|system-source|stardate|snmptrapd-parser|xml|openbsd|examples|kafka|afsnmp|mqtt|regexp-parser|[^/]*$)
```
, and the last check was successful. And it seems that the console output of `Kira-starter` doesn't explain why it failed.
Any help will be appreciated. Thank you.

Kokan commented:
@LittleFish33 I've checked, the Kira failure is not due to your changes. Actually there are some cases on master that's incorrect. You cannot and shouldn't do anything regarding that, once fixed we'll notify you.

Kokan commented:
@kira-syslogng retest this please

kira-syslogng commented:
Build SUCCESS

mitzkia commented:
@kira-syslogng test this please;

kira-syslogng commented:
Build SUCCESS

mitzkia commented on tests/python_functional/functional_tests/parsers/regexp-parser/test_regexp_parser.py:
```diff
@@ -0,0 +1,56 @@
+
+
+test_parameters_raw = [
+    ("foo", '''""''', '''"(?<DN>foo)"''', "", "${DN}", True, "foo"),
```
First of all thanks for the Light testcases.
My comment for different testcases: 
- Can you please add a more basic testcase, when 'patterns' is a simple pattern match without any reference saving: \<DN\>. In this case you can use a custom prefix value to store the result.
- Can you please give a better name for DN. If I am right it is just custom macro name
- Could you please give a testcase for this statement also? "use multiple regular expressions, these should be processed in order, the first successfully matching one causes the processing to stop"
- Could you please add a testcase for the following: "Error parsing regexp-parser, compiling the regexp failed", in this case syslog-ng should not started

LittleFish33 replied:
Thank you for your comments, I will work on this soon.

LittleFish33 replied:
I have updated the commits with the following changes:
* A new test case has been added, whose `id` is `match_without_reference_saving`
* I change the reference name to `<key>`
* A new test case has been added, whose `id` is `match_multiple_regular_expressions`
* A new test case has been added, whose `id` is `regular_expression_compile_error`
* A new test case has been added, whose `id` is `check_MSG_macro`
* A new test case has been added, whose `id` is `unmatch_case`. I have made it available to use 'DriverStatsHandler' in Parser class and extended the API in 'DriverStatsHandler' for parsers. The details can be found in commits [a16d34](https://github.com/syslog-ng/syslog-ng/pull/3702/commits/a16d34f3e3fef3840c040bab565834b3fad34925) and [4b2b90](https://github.com/syslog-ng/syslog-ng/pull/3702/commits/4b2b90a2548537a690cd0dba78e06b6bb7634334)

mitzkia replied:
Thanks for the changes, I will check it soon.

mitzkia replied:
Thanks for the changes again, they are very good. I have only a few minor review comments.

mitzkia commented on tests/python_functional/functional_tests/parsers/regexp-parser/test_regexp_parser.py:
```diff
@@ -0,0 +1,56 @@
+    ids=list(map(str, range(len(test_parameters_raw)))),
+)
+def test_regexp_parser(config, syslog_ng, input_message, prefix, patterns, template, flags, expected_result, expected_value):
+    config.add_include("scl.conf")
```
This testcase does not require scl.conf

LittleFish33 replied:
`config.add_include("scl.conf")` has been removed.

bazsi commented on modules/regexp-parser/regexp-parser.c:
```diff
@@ -0,0 +1,178 @@
+      self->matchers = g_list_prepend(self->matchers, log_matcher_new(&self->matcher_options));
+      log_matcher_pcre_set_prefix((LogMatcher *)self->matchers->data, self->prefix);
+
+      if(!log_matcher_compile((LogMatcher *)self->matchers->data, item->data, NULL))
```
Please handle the error in a more user-friendly way, I got no reason why the compilation of my regexp failed:

Error parsing regexp-parser, compiling the regexp failed in etc/syslog-ng-re.conf:5:11-5:61:
1       @version: 3.32
2       
3       log { 
4       	source { tcp(port(2000)); };
5-----> 	parser { regexp-parser(patterns("(?foo)a+") prefix(".re.")); };
5-----> 	         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
6       	destination { file("/tmp/log" template("$(format-json *)\n")); };
7       };

LittleFish33 replied:
I have updated the `regexp-parser-grammar.ym` file and the `regexp_parser_compile` function. In the latest commits, syslog-ng will output the error with the failure reason:
```
Error parsing regexp-parser, compiling the regexp failed, error=Failed to compile PCRE expression >>>(?foo)a+<<< `unrecognized character after (? or (?-' at character 2 in /etc/syslog-ng.conf:57:11-57:61:
55      log { 
56      	source(s_file);
57----> 	parser { regexp-parser(patterns("(?foo)a+") prefix(".re.")); };
57----> 	         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
58      	destination { file("/tmp/log" template("$(format-json *)\n")); };
59      };
```

bazsi commented:
While testing the regexp-parser() plugin, I've encountered that it applies all regexps to the input string, regardless if they matched or not, the original GSoC specification contained this point: 

> use multiple regular expressions, these should be processed in order, the first successfully matching one causes the  processing to stop

I was using this config:
```
@version: 3.32

log { 
	source { tcp(port(2000)); };
	parser { regexp-parser(patterns("(?<foo>)a+", "(?<bar>)b+") prefix(".re.")); };
	destination { file("/tmp/log" template("$(format-json *)\n")); };
};
```
Albeit applying all of them could make sense in some cases, the general idea is that you specify multiple regexps to support multiple incarnations of the _same_ meaning, potentially in some variations. In this case, it is expected that you list the alternative regexps to parse the same thing. And in this case it would make sense to quickly break out of the matching loop the first time it matches, expecting the patterns to be ordered by their likelyhood of being encountered.

Please either break out at the first match or add an option to short-circuit at the first matching regexp and make it the default behavior.

thanks

bazsi commented:
I have resolved most of the open questions, a few still remained. I manually tested the functionality and it works great. I have one issue pointed out in one of my latest comments, but it's very close to become merge-able.

@alltilla could you also make a review so we have two reviews to get it in?

Could you fix the remaining issues?

thanks

alltilla commented on modules/regexp-parser/regexp-parser-grammar.ym:
```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright (c) 2021 Balabit
```
Please change "Balabit" to "One Identity" here, and in every file you have created.

LittleFish33 replied:
Related changes have been udpated.

alltilla commented on modules/regexp-parser/regexp-parser.c:
```diff
@@ -0,0 +1,178 @@
+
+  g_free(self->prefix);
+  string_list_free(self->patterns);
+  log_parser_free_method(s);
```
Please make this the last call in this function.

LittleFish33 replied:
Related changes have been udpated.

alltilla reviewed (commented):
I agree with Bazsi about the early return in `process()`. Please also add unit tests that make sure we do not store matches after the first matching pattern.

---

Please add a news file entry. See `news/README.md`.

---

Please make the commit message(s) more verbose.
Tell about the new module, and show an example config snippet.

---

Also it would have been better from a reviewer's and maintainer's perspective, if this code change was split into some smaller commits:
 * Adding `prefix` to `LogMatcher`.
 * Adding the skeleton of the new module `regexp-parser`.
 * Adding the implementation of `regexp-parser`.
 * Adding unit tests.
 * Adding functional tests.
 * Adding `mod-regexp-parser` to the packages.
 * Adding news file.

Please do so.

kira-syslogng commented:
Build SUCCESS

LittleFish33 commented on tests/python_functional/functional_tests/parsers/regexp-parser/test_regexp_parser.py:
```diff
@@ -0,0 +1,56 @@
+
+@pytest.mark.parametrize(
+    "input_message, prefix, patterns, flags, template, expected_result, expected_value", test_parameters_raw,
+    ids=list(map(str, range(len(test_parameters_raw)))),
```
Thanks, I have added `ids` for each test case in the latest commits.

LittleFish33 commented on lib/logmatcher.c:
```diff
@@ -291,6 +292,8 @@ typedef struct _LogMatcherPcreRe
   pcre_extra *extra;
   gint match_options;
+  gchar *prefix;
+  gint prefix_len;
```
I have changed the `prefix` and `prefix_len` to `nv_prefix` and `nv_prefix_len`.

LittleFish33 commented:
> I agree with Bazsi about the early return in `process()`. Please also add unit tests that make sure we do not store matches after the first matching.

I have updated the code and added a new test case for this change. The test case `id` is `match_multiple_regular_expressions`.

bazsi commented:
Thanks for the updates. I will check them out. Hopefully this is the last
iteration.

On Sun, Jul 18, 2021, 07:30 LittleFish ***@***.***> wrote:

> @LittleFish33 <https://github.com/LittleFish33> requested your review on:
> #3702 <https://github.com/syslog-ng/syslog-ng/pull/3702> GSoC 2021
> Project: Add regexp-parser().
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/syslog-ng/syslog-ng/pull/3702#event-5034688593>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAFOK5RJ5H2CV7S6E4QYCQTTYJRHXANCNFSM46S7ZNWQ>
> .
>

alltilla reviewed (dismissed):
I only have 1 optional review note. Approve.

kira-syslogng commented:
Build SUCCESS

LittleFish33 commented on modules/regexp-parser/regexp-parser.c:
```diff
@@ -0,0 +1,175 @@
+      if ((G_LIKELY(!self->super.template)
+           && log_matcher_match((LogMatcher *)item->data, *pmsg, LM_V_MESSAGE, input, input_len)) ||
+          (!G_LIKELY(!self->super.template)
+           && log_matcher_match((LogMatcher *)item->data, *pmsg, LM_V_NONE, input, input_len)))
```
Thank you for your comments. I've updated the code as suggested.

alltilla reviewed (approved):
Thanks!

mitzkia commented on tests/python_functional/functional_tests/parsers/regexp-parser/test_regexp_parser.py:
```diff
@@ -0,0 +1,67 @@
+        "check_MSG_macro",
+    ],
+)
+def test_regexp_parser(config, syslog_ng, syslog_ng_ctl, input_message, prefix, patterns, flags, template, compile_result, expected_result, expected_value):
```
Thanks for the changes, they are very good.
My comment here: syslog_ng_ctl fixture is not used in the testcase, it can be deleted from testcase parameter list

mitzkia commented on tests/python_functional/functional_tests/parsers/regexp-parser/test_regexp_parser.py:
```diff
@@ -0,0 +1,67 @@
+    file_destination = config.create_file_destination(file_name="output.log", template=config.stringify(template + "\n"))
+    config.create_logpath(statements=[generator_source, regexp_parser, file_destination])
+
+    if compile_result is True:
```
It is enough to use: 'if compile_result:'

mitzkia commented on tests/python_functional/functional_tests/parsers/regexp-parser/test_regexp_parser.py:
```diff
@@ -0,0 +1,67 @@
+    if compile_result is True:
+        syslog_ng.start(config)
+        if expected_result:
+            assert file_destination.read_log().strip() == expected_value
```
Maybe we can add 'assert regexp_parser.get_query() == {'discarded': 0}' line after 'assert file_destination.read_log().strip() == expected_value'

LittleFish33 replied:
Thanks for your comments, I've updated the code.

mitzkia reviewed (commented):
Thanks for the changes. 
Other than my minor review comments I am approving functional testcases for Light

kira-syslogng commented:
Build SUCCESS

mitzkia commented:
Thanks for the functional testcases.
From my point of view functional Light testcases are approved.

bazsi reviewed (approved):
After doing the review iterations and fixing the notes (thank you!) I am happy with this PR, so approve from me.
